src/java.base/share/classes/java/net/URLEncoder.java

```diff
@@ -268,23 +275,47 @@ public static String encode(String s, Charset charset) {
                         if ((i + 1) < s.length()) {
                             char d = s.charAt(i + 1);
                             if (Character.isLowSurrogate(d)) {
-                                charArrayWriter.write(d);
+                                cb.put(d);
                                 i++;
                             }
                         }
                     }
+                    // Limit to ENCODING_CHUNK_SIZE - 1 so that we can always fit in
+                    // a surrogate pair on the next iteration
+                    if (cb.position() >= ENCODING_CHUNK_SIZE - 1) {
+                        flushToStringBuilder(out, ce, cb, bb, false);
+                    }
                     i++;
                 } while (i < s.length() && !DONT_NEED_ENCODING.test((c = s.charAt(i))));
-
-                String str = charArrayWriter.toString();
-                byte[] ba = str.getBytes(charset);
-                for (byte b : ba) {
-                    encodeByte(out, b);
-                }
-                charArrayWriter.reset();
+                flushToStringBuilder(out, ce, cb, bb, true);
             }
         }
-
         return out.toString();
     }
+
+    private static void flushToStringBuilder(StringBuilder out, CharsetEncoder ce, CharBuffer cb, ByteBuffer bb, boolean endOfInput) {
```

Comment:
This method would benefit from a comment to summarise the parameters, only because there are so many. 

Also just a formatting nit is that this line is much longer than everything else in the file so maybe side-by-side diffs a bit difficult to look at.

Reply:
Added a comment and split up the line.